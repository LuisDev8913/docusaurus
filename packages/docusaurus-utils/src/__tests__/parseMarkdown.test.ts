/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {parseMarkdownString, readFrontMatter} from '../index';
import dedent from 'dedent';

describe('load utils: parseMarkdown', () => {
  describe('readFrontMatter', () => {
    test('should read front matter', () => {
      const warn = jest.spyOn(console, 'warn').mockImplementation(() => {});
      expect(
        readFrontMatter(dedent`
        ---
        title: test
        ---
        `),
      ).toMatchSnapshot();
      expect(warn).not.toBeCalled();
    });
    test('should parse first heading as title', () => {
      const warn = jest.spyOn(console, 'warn').mockImplementation(() => {});
      expect(
        readFrontMatter(dedent`
        # test
        `),
      ).toMatchSnapshot();
      expect(warn).not.toBeCalled();
    });
    test('should preserve front-matter title and warn about duplication', () => {
      const warn = jest.spyOn(console, 'warn').mockImplementation(() => {});
      expect(
        readFrontMatter(dedent`
        ---
        title: title
        ---
        # test
        `),
      ).toMatchSnapshot();
      expect(warn).toBeCalledWith('Duplicate title detected in `this` file');
      warn.mockReset();
    });
    test('should ignore heading if its not a first text', () => {
      const warn = jest.spyOn(console, 'warn').mockImplementation(() => {});
      expect(
        readFrontMatter(dedent`
        foo
        # test
        `),
      ).toMatchSnapshot();
      expect(warn).not.toBeCalled();
    });
    test('should parse first heading as title and keep it in content', () => {
      const warn = jest.spyOn(console, 'warn').mockImplementation(() => {});
      expect(
        readFrontMatter(
          dedent`
          # test
          `,
          undefined,
          {},
          false,
        ),
      ).toMatchSnapshot();
      expect(warn).not.toBeCalled();
    });
    test('should delete only first heading', () => {
      const warn = jest.spyOn(console, 'warn').mockImplementation(() => {});
      expect(
        readFrontMatter(dedent`
        # test
        test test test # test bar
        # test
        ### test
        `),
      ).toMatchSnapshot();
      expect(warn).not.toBeCalled();
    });
    test('should parse front-matter and ignore h2', () => {
      const warn = jest.spyOn(console, 'warn').mockImplementation(() => {});
      expect(
        readFrontMatter(
          dedent`
          ---
          title: title
          ---
          ## test
          `,
          undefined,
          {},
          false,
        ),
      ).toMatchSnapshot();
      expect(warn).not.toBeCalled();
    });
    test('should not warn about duplicated title', () => {
      const warn = jest.spyOn(console, 'warn').mockImplementation(() => {});
      expect(
        readFrontMatter(
          dedent`
          ---
          title: title
          ---
          # test
          `,
          undefined,
          {},
          false,
        ),
      ).toMatchSnapshot();
      expect(warn).not.toBeCalled();
    });
  });

  describe('parseMarkdownString', () => {
    test('should read front matter', () => {
      const warn = jest.spyOn(console, 'warn').mockImplementation(() => {});
      expect(
        parseMarkdownString(dedent`
        ---
        title: test
        ---
        `),
      ).toMatchSnapshot();
      expect(warn).not.toBeCalled();
    });
    test('should parse first heading as title', () => {
      const warn = jest.spyOn(console, 'warn').mockImplementation(() => {});
      expect(
        parseMarkdownString(dedent`
        # test
        `),
      ).toMatchSnapshot();
      expect(warn).not.toBeCalled();
    });
    test('should preserve front-matter title and warn about duplication', () => {
      const warn = jest.spyOn(console, 'warn').mockImplementation(() => {});
      expect(
        parseMarkdownString(dedent`
        ---
        title: title
        ---
        # test
        `),
      ).toMatchSnapshot();
      expect(warn).toBeCalledWith('Duplicate title detected in `this` file');
      warn.mockReset();
    });
    test('should ignore heading if its not a first text', () => {
      const warn = jest.spyOn(console, 'warn').mockImplementation(() => {});
      expect(
        parseMarkdownString(dedent`
        foo
        # test
        `),
      ).toMatchSnapshot();
      expect(warn).not.toBeCalled();
    });
    test('should delete only first heading', () => {
      const warn = jest.spyOn(console, 'warn').mockImplementation(() => {});
      expect(
        parseMarkdownString(dedent`
        # test

        test test test test test test
        test test test # test bar
        # test
        ### test
        `),
      ).toMatchSnapshot();
      expect(warn).not.toBeCalled();
    });
  });
});
