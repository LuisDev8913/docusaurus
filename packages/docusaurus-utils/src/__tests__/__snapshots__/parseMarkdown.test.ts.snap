// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load utils: parseMarkdown parseMarkdownString should delete only first heading 1`] = `
Object {
  "content": "
test test test test test test
test test test # test bar
# test
### test",
  "excerpt": "",
  "frontMatter": Object {
    "title": "test",
  },
  "hasFrontMatter": false,
}
`;

exports[`load utils: parseMarkdown parseMarkdownString should ignore heading if its not a first text 1`] = `
Object {
  "content": "foo
# test",
  "excerpt": "foo",
  "frontMatter": Object {},
  "hasFrontMatter": false,
}
`;

exports[`load utils: parseMarkdown parseMarkdownString should parse first heading as title 1`] = `
Object {
  "content": "",
  "excerpt": "",
  "frontMatter": Object {
    "title": "test",
  },
  "hasFrontMatter": false,
}
`;

exports[`load utils: parseMarkdown parseMarkdownString should preserve front-matter title and warn about duplication 1`] = `
Object {
  "content": "# test",
  "excerpt": "test",
  "frontMatter": Object {
    "title": "title",
  },
  "hasFrontMatter": true,
}
`;

exports[`load utils: parseMarkdown parseMarkdownString should read front matter 1`] = `
Object {
  "content": "",
  "excerpt": undefined,
  "frontMatter": Object {
    "title": "test",
  },
  "hasFrontMatter": true,
}
`;

exports[`load utils: parseMarkdown readFrontMatter should delete only first heading 1`] = `
Object {
  "content": "test test test # test bar
# test
### test",
  "excerpt": "",
  "frontMatter": Object {
    "title": "test",
  },
  "hasFrontMatter": false,
}
`;

exports[`load utils: parseMarkdown readFrontMatter should ignore heading if its not a first text 1`] = `
Object {
  "content": "foo
# test",
  "excerpt": "",
  "frontMatter": Object {},
  "hasFrontMatter": false,
}
`;

exports[`load utils: parseMarkdown readFrontMatter should parse first heading as title 1`] = `
Object {
  "content": "",
  "excerpt": "",
  "frontMatter": Object {
    "title": "test",
  },
  "hasFrontMatter": false,
}
`;

exports[`load utils: parseMarkdown readFrontMatter should parse first heading as title and keep it in content 1`] = `
Object {
  "content": "# test",
  "excerpt": "",
  "frontMatter": Object {
    "title": "test",
  },
  "hasFrontMatter": false,
}
`;

exports[`load utils: parseMarkdown readFrontMatter should parse front-matter and ignore h2 1`] = `
Object {
  "content": "## test",
  "excerpt": "",
  "frontMatter": Object {
    "title": "title",
  },
  "hasFrontMatter": true,
}
`;

exports[`load utils: parseMarkdown readFrontMatter should preserve front-matter title and warn about duplication 1`] = `
Object {
  "content": "# test",
  "excerpt": "",
  "frontMatter": Object {
    "title": "title",
  },
  "hasFrontMatter": true,
}
`;

exports[`load utils: parseMarkdown readFrontMatter should read front matter 1`] = `
Object {
  "content": "",
  "excerpt": "",
  "frontMatter": Object {
    "title": "test",
  },
  "hasFrontMatter": true,
}
`;
